File: POH6/rio/main.go
//
// 結果 https://paiza.jp/poh/joshibato/rio/result/0edbfb23
//
package main
import "fmt"

func main() {
    
    var n int
    fmt.Scanf("%d", &n)
    
    var w float64 = 0.0
    var c float64 = 0.0
    
    for i := 0; i < n; i++ {
        var t int
        var s float64
        fmt.Scanf("%d %f", &t, &s)
        switch t {
        case 1:
            w += s
        case 2:
            c += s
        case 3:
            wc := w + c
            w = (w * wc - s * w) / wc
            c = (c * wc - s * c) / wc
        }
        
    }

    ans := 100.0 * c / (w + c)
    fmt.Printf("%d\n", int(ans))
}

File: POH6/rio/Main.scala
/*
 * 結果 https://paiza.jp/poh/joshibato/rio/result/c24a94c3
 */
object Main extends App{
    
    def solve(n: Int, water: Double, coffee: Double): Unit = {
        if (n == 0) {
            val ans = 100.0 * coffee / (water + coffee);
            println(ans.intValue());
        } else {
            val tas = scala.io.StdIn.readLine().split(" ")
            val s = tas(1).toDouble
            tas(0) match {
            case "1" => solve(n - 1, water + s, coffee)
            case "2" => solve(n - 1, water, coffee + s)
            case  _  => val wc = water + coffee
                        solve(n - 1, (water * wc - s * water) / wc, (coffee * wc - s * coffee) /  wc)
            }
        }
    }
    
    val n = scala.io.StdIn.readLine().toInt
    
    solve(n, 0.0, 0.0)
}
